From 969881200bb6971ef690f1d476db49021f5d71d2 Mon Sep 17 00:00:00 2001
From: Arulanbu Balusamy <quic_abalusam@quicinc.com>
Date: Fri, 1 Dec 2023 15:59:41 +0530
Subject: [PATCH] wifi: ath12k: fix tx power, max reg power update to firmware

Currently values for max_regpower, max_power, max_antenna_gain are
multiplied by 2 before sending to firmware which is wrong since firmware uses
1dBm step for power calculations.

This leads to wrong power values used in firmware and radio.

Fix the update of max_regpower, max_power, max_antenna_gain values
and also add missing max tx power update in SCAN CHAN LIST WMI cmd.

Signed-off-by: Arulanbu Balusamy <quic_abalusam@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 12 ++++++------
 drivers/net/wireless/ath/ath12k/wmi.c |  2 ++
 2 files changed, 8 insertions(+), 6 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -11546,9 +11546,9 @@
 		ath12k_phymodes[chandef->chan->band][chandef->width];
 
 	arg.channel.min_power = 0;
-	arg.channel.max_power = chandef->chan->max_power * 2;
-	arg.channel.max_reg_power = chandef->chan->max_reg_power * 2;
-	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain * 2;
+	arg.channel.max_power = chandef->chan->max_power;
+	arg.channel.max_reg_power = chandef->chan->max_reg_power;
+	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain;
 	arg.ru_punct_bitmap = ~chandef->ru_punct_bitmap;
 
 	arg.pref_tx_streams = ar->num_tx_chains;
@@ -12168,9 +12168,9 @@
 		ath12k_phymodes[chandef->chan->band][chandef->width];
 
 	arg.channel.min_power = 0;
-	arg.channel.max_power = chandef->chan->max_power * 2;
-	arg.channel.max_reg_power = chandef->chan->max_reg_power * 2;
-	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain * 2;
+	arg.channel.max_power = chandef->chan->max_power;
+	arg.channel.max_reg_power = chandef->chan->max_reg_power;
+	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain;
 	arg.channel.chan_radar = !!(chandef->chan->flags & IEEE80211_CHAN_RADAR);
 	arg.channel.passive = arg.channel.chan_radar;
 	arg.channel.freq2_radar = radar_enabled;
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -3449,6 +3449,8 @@
 						 WMI_CHAN_REG_INFO1_REG_CLS);
 			*reg2 |= u32_encode_bits(tchan_info->antennamax,
 						 WMI_CHAN_REG_INFO2_ANT_MAX);
+			*reg2 |= u32_encode_bits(tchan_info->maxregpower,
+						WMI_CHAN_REG_INFO2_MAX_TX_PWR);
 
 			ath12k_dbg(ar->ab, ATH12K_DBG_WMI,
 				   "WMI chan scan list chan[%d] = %u, chan_info->info %8x\n",
